src/sinks/datadog/logs/sink.rs

```diff
@@ -0,0 +1,429 @@
+        assert!(arr.len() + 1 <= MAX_PAYLOAD_ARRAY);
+        arr.push(event);
+    }
+
+    /// Flush the `Event` batches out to Datadog API
+    async fn flush_to_api<'a>(
+        self: &'a mut Self,
+    ) -> Result<impl Iterator<Item = Client::Future> + Send + 'a, ()> {
+        // The Datadog API makes no claim on how many requests we can make at
+        // one time. We rely on the http client to handle retries and what
+        // not. This implementation stacks up futures for each key present in
+        // the batches and runs them unordered, returning when the last
+        // completes.
+
+        // NOTE I would prefer to map directly over the Drain without going
+        // through a vec but mixing ownership confused me greatly, since we need
+        // to call back to `self` to flush the batch.
+        let batches: Vec<(u64, Vec<Event>)> = self.event_batches.drain().collect();
+        poll_fn(|cx| self.http_client.poll_ready(cx))
+            .await
+            .map_err(|_e| ())?;
+        let futures = // : dyn Iterator<Item = Client::Future> =
+            batches.into_iter().map(move |(api_key_id, batch)| {
+                let members: Vec<BTreeMap<String, Value>> = batch
+                    .into_iter()
+                    .map(|event| event.into_log())
+                    .map(|log| log.into_parts().0)
+                    .collect();
+                let payload = Payload { members };
+                let body: Vec<u8> = serde_json::to_vec(&payload).expect("failed to encode to json");
+
+                let api_key = self.key_slab.get(&api_key_id).expect("impossible situation");
+                let request = Request::post(self.datadog_uri.clone())
+                    .header("Content-Type", "application/json")
+                    .header("DD-API-KEY", &api_key[..]);
+                let (request, encoded_body) = match self.compression {
+                    Compression::None => (request, body),
+                    Compression::Gzip(level) => {
+                        let level = level.unwrap_or(GZIP_FAST);
+                        let mut encoder = GzEncoder::new(
+                            Vec::with_capacity(body.len()),
+                            flate2::Compression::new(level as u32),
+                        );
+
+                        encoder.write_all(&body).expect("failed to write body");
+                        (
+                            request.header("Content-Encoding", "gzip"),
+                            encoder.finish().expect("failed to encode"),
+                        )
+                    }
+                };
+                let request = request
+                    .header("Content-Length", encoded_body.len())
+                    .body(Body::from(encoded_body))
+                    // .map_err(Into::into)
+                    .expect("failed to make request");
+
+                self.http_client.call(request)
```

Comment:
Then, after this call resolves, you would need catch the result of the call and call `update_status`, something like this:
```suggestion
                async move {
                    let result = self.http_client.call(request).await;
                    let status: EventStatus = match result { … };
                    for finalizer in finalizers {
                        finalizer.update_status(status);
                    }
                    result
                }
```